046-analyse-broker-metrics-before-rolling-kafka.md

```diff
@@ -0,0 +1,86 @@
+# Analyse broker metrics before rolling Kafka
+
+This proposal describes a solution for KafkaRoller's shortcoming with brokers in log recovery.
+
+## Current situation
+
+The current logic in KafkaRoller:
+- Take a list of Kafka pods and reorder them based on the readiness.
+- Roll unready pods first without considering why it could be unready.
+- Before restarting a Kafka pod, consider if it is a controller or if it has an impact on the availability.
+- If a Kafka pod does not become ready within the operational timeout (300000ms by default) after restarting, then force restart it without considering why it's taking a long time.
+
+A Kafka broker can take a long time to become ready while performing log recovery. In this case, KafkaRoller ends up force restarting the broker and this can continue indefinitely, because of the log recovery on the broker startup again. KafkaRoller currently does not have a way to know what state a broker is in before making a decision to force restart it.
+ 
+KafkaAgent currently collects the `kafka.server:type=KafkaServer,name=BrokerState` metric and creates a `kafka-ready` file on disk if the broker state value is `3`(`RUNNING`). The Kafka readiness check passes if it finds the file on disk.
+
+## Motivation
+
+To address the issues with restarts and log recovery, KafkaRoller needs to be aware of the current broker state.  If a broker is performing log recovery, KafkaRoller should fail the reconciliation instead of force restarting it.
+
+## Proposal
+ 
+The objective of this proposal is to introduce a REST API endpoint within KafkaAgent, which exposes metrics from a Kafka broker. Furthermore, KafkaRoller would be modified so that it can query this endpoint to determine the current state of a broker, thus preventing unnecessary restarts. Additionally, KafkaRoller would also display the remaining logs and segments to recover, allowing a human operator to observe the recovery process and determine if it is making progress or is stuck.
+
+### Implementation details
+
+*KafkaAgent*:
+- Add a web server to accept SSL connections on port 8443.
+- Use Jetty for the web server since Kafka already uses it therefore avoids an additional dependency.
+- Reuse the broker's own certificate that would be in the container for TLS authentication. 
+- Configure the server to require TLS client authentication. 
+- Expose an endpoint `v1/brokermetrics` that includes the API version in the URI path.
```

Comment:
This is good. But I wonder if you need discovery as well => i.e. way to find out which APIs are supported. So that you don't have to blindly try the versions in the future in some upgrade. E.g. some `apis/` endpoint which would return a list of supported API versions ... e.g. just `["v1"]` right now. That would allow the future rollers to decide what it is talking with and use the right API during upgrades. My guess is that this might be useful one day and it would not cost much on the implementation effort (no need to any sophisticated mechanisms in the KAfkaAgent, just return a static data 🤷 ).

Reply:
I wonder if that is really better:

* `GET /v2/brokermetrics` => 404, `GET /v1/brokermetrics` => 200
* `GET /apis` => 200, `GET /v1/brokermetrics` => 200

It's two `GET` requests in both cases, but `/apis` means we need a separate endpoint on the server side,  to make a separate kind of request on the client side, and brings a requirement to model endpoints for the response (extra serialization and deserialization, even if it's trivial). I think I'd be inclined to keep it simple for now.

Reply:
Well, maybe I'm a bit pessimistic about the versioning ... but you can also imagine it like this 😉:
* GET /v5/brokermetrics => 404
* GET /v4/brokermetrics => 404
* GET /v3/brokermetrics => 404
* GET /v2/brokermetrics => 404
* GET /v1/brokermetrics => 200

As we need to support upgrades across versions, it is not necessarily just the current version and the previous one. You might need to support more of them.

Reply:
But the second case will always be two `GET`(s) (getting the API version and then asking for it).
The first one could be more if the roller starts to try with the latest one, maybe `v4` but ends after 3 times of 404 to get 200 for `v1`. Or it could not happen to have a so updated roller communicating with a potential so old agent?

Reply:
When we got to that point I think we'd have the justification to consider `/apis`. And to be honest I don't see the evolution of this endpoint being so rapid that going that far back in the versions was at all common (i.e. because operator and broker images tend to be upgraded together).

Reply:
@scholzj I was with you, just clicking "submit review" later :-)

Reply:
As majority agreed, keeping it simple for now and going with 404 error for invalid version. We can implement something smarter to handle the versions when we have multiple of them.

Reply:
> When we got to that point I think we'd have the justification to consider /apis. And to be honest I don't see the evolution of this endpoint being so rapid that going that far back in the versions was at all common (i.e. because operator and broker images tend to be upgraded together).

The point is that you cannot do it _when it comes to it_. Because then it would not be in the previous versions and you will not be able to rely on it. And rolling it out when we get to `v5` would mean for example waiting for `v10` before you can make any use for it. So it is not something _we do when we get there_. You cannot really do it when you get there.

The point of this is also not the regular rolling updates but upgrades. So regardless of `operator and broker images tend to be upgraded together`, you will need to deal with these situations when the operator from Strimzi 0.44 will be doing the upgrade from Strimzi 0.34 for example.

I do not really know how quickly will the API evolve. Maybe we will still use `v1` in 2 years. But who knows?

Reply:
I mean, if we want to stick with the 404, then fine, I can deal with it. But the reasoning from @tombentley is not valid and does not make sense. That is what I'm trying to point out. When we are at v5, adding the `/apis` endpoint would not help with v1-v5. And just because we are at `v5`, the are is no guarantee that we will add more versions in the future as we maybe just found the right version. So the logic there is wrong.